src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -46,18 +50,93 @@ protected virtual CompletionItemRules GetCompletionItemRules(IReadOnlyList<ISymb
         private ImmutableArray<CompletionItem> CreateItems(
             ImmutableArray<ISymbol> symbols,
             SyntaxContext context,
-            bool preselect)
+            bool preselect,
+            ImmutableArray<ITypeSymbol> inferredTypes = default)
         {
-            var tree = context.SyntaxTree;
+            if (IsTargetTypeCompletionFilterExperimentEnabled(context.Workspace))
+            {
+                var symbolGroups = from symbol in symbols
+                                   let texts = GetDisplayAndSuffixAndInsertionText(symbol, context)
+                                   group symbol by texts into g
+                                   select g;
 
-            var q = from symbol in symbols
-                    let texts = GetDisplayAndSuffixAndInsertionText(symbol, context)
-                    group symbol by texts into g
-                    select this.CreateItem(
-                        g.Key.displayText, g.Key.suffix, g.Key.insertionText, g.ToList(), context,
-                        invalidProjectMap: null, totalProjects: null, preselect: preselect);
+                var itemListBuilder = ImmutableArray.CreateBuilder<CompletionItem>();
 
-            return q.ToImmutableArray();
+                foreach (var symbolGroup in symbolGroups)
+                {
+                    var item = this.CreateItem(
+                            symbolGroup.Key.displayText, symbolGroup.Key.suffix, symbolGroup.Key.insertionText, symbolGroup.ToList(), context,
+                            invalidProjectMap: null, totalProjects: null, preselect: preselect);
+
+                    if (IsTargetTypeCompletionFilterExperimentEnabled(context.Workspace))
```

Comment:
why the check outside and inside for this?

Reply:
Because refactoring is hard. Will fix.